components/tikv_util/src/trend.rs

```diff
@@ -0,0 +1,737 @@
+            return Some(vec![curr]);
+        }
+
+        // In a spike
+
+        // Spike ending check
+        if (value as f64) < exit_threshold {
+            if self.values.len() <= 2 {
+                // The checking sequence is too short to be a spike, dump all and exit checking
+                let mut values: Vec<SampleValue> = self.values.drain(..).collect();
+                values.push(curr);
+                return Some(values);
+            }
+            // The checking sequence is not long enough to be regular high, it's a spike,
+            // discard all but return curr
+            self.filter_value_gauge.set(self.avg() as i64);
+            self.filter_count_gauge.inc();
+            self.values.drain(..);
+            return Some(vec![curr]);
+        }
+
+        // Hold curr value to this spike
+        self.values.push_back(curr);
+        None
+    }
+
+    #[inline]
+    fn avg(&self) -> f64 {
+        if self.values.is_empty() {
+            return 0.0;
+        }
+        let mut sum: f64 = 0.0;
+        for value in self.values.iter() {
+            sum += value.value as f64;
+        }
+        sum / (self.values.len() as f64)
+    }
+}
+
+// Responsibilities of each window:
+//
+// L0:
+//     Eleminate very short time jitter,
+//     Consider its avg value as a point in data flow
+// L1:
+//     `L0.avg/L1.avg` to trigger slow-event, not last long but high sensitive
+//     Sensitive could be tuned by `L0.duration` and `L1.duration`
+//     Include periodic fluctuations, so it's avg could be seen as baseline
+// value     Its duration is also the no-detectable duration after TiKV starting
+// L2:
+//     `L1.avg/L2.avg` to trigger slow-event, last long but low sensitive
+//     Sensitive could be tuned by `L1.duration` and `L2.duration`
+//
+// L* History:
+//     Sample history values and calculate the margin error
+//
+// Spike Filter:
+//     Erase very high and short time spike-values
+//
+pub struct Trend {
+    sample_interval: usize,
+    sample_sequence_id: usize,
+
+    spike_filter: SpikeFilter,
+    spike_filter_enabled: bool,
+
+    data_flow: SampleWindows,
+
+    l1_history: HistoryWindow,
+    l2_history: HistoryWindow,
+
+    // When SPOT is being used, these should be `risk multiple`
+    l1_margin_error_multiple: f64,
+    l2_margin_error_multiple: f64,
+
+    curves_composer: CurvesComposer,
+}
+
+impl Trend {
+    pub fn new(
+        spike_filter_duration: Duration,
+        spike_filter_value_gauge: IntGauge,
+        spike_filter_count_gauge: IntGauge,
+        history_duration: Duration,
+        l0_duration: Duration,
+        l1_duration: Duration,
+        l2_duration: Duration,
+        sample_interval: usize,
+        tolerable_margin_error_value: u64,
+        l1_gap_gauge: IntGauge,
+        l2_gap_gauge: IntGauge,
+        unsensitive_multiple: f64,
+    ) -> Self {
+        let margin_error_base = tolerable_margin_error_value as f64;
+        Self {
```

Comment:
The `tolerable_margin_error_value` will be automatically replaced after the windows got enough data (~10min), so as long as it is small enough (if it's too big then it won't be covered), then its value is rather trivial/irrelevant.

For the bunch value of `xxx_multiple`, we could get very good result without them, but with them could reduce the chances of `false alarm` (even the chances are tiny, but still, the less the better).

The cost of high (as the current code) `xxx_multiple` is a bit lower sensitiveness, here is an example by the test result:
* When `unsensitive_multiple` for `cause trend` was 2, we got almost none `false alarm` in the `cause trend` curve (a very tiny bit after days' tests, no big deal because the `cause-result` double checking handled it well so no wrong evicting happened)
* When `unsensitive_multiple` for `cause trend` was 10 (current value), none `false alarm` happened, the detecting sensitiveness lower from `<30s` to `<45s`, not changed much, so it's worth it to increase the threshold to make the detecting safer

Could we get rid of those in the future (when we use SPOT)? Yes and No:
* Yes: all the `xxx_multiple` values like `2` `3` `5` will be replaced by `risk` params like `99.98%`, it will be much more safer and easier
* No: `99.98%` or ``99.95%`, that could still make some differencies, so the pre-tune (in the developing stage) work might still needed